lib/subgraphManagers/IntrospectAddNodeAndCompose.ts

```diff
@@ -0,0 +1,321 @@
+    this.config = options;
+    this.state = { phase: 'initialized' };
+  }
+
+  public async initialize({
+    update,
+    getDataSource,
+    healthCheck,
+  }: SupergraphSdlHookOptions): Promise<{
+    supergraphSdl: string;
+    cleanup: () => Promise<void>;
+  }> {
+    this.update = update;
+
+    if (this.config.subgraphHealthCheck) {
+      this.healthCheck = healthCheck;
+    }
+    this.getDataSource = getDataSource;
+
+    this.subgraphs = this.config.subgraphs.map(subgraph => ({
+      ...subgraph,
+      dataSource: getDataSource(subgraph),
+    }));
+
+    this.originalSubgraphs = [...this.subgraphs];
+
+    let initialSupergraphSdl: string | null =
+      this.completeSupergraphSdl || null;
+    try {
+      initialSupergraphSdl = await this.updateSupergraphSdl().then(sdl => {
+        if (sdl) {
+          return sdl;
+        }
+        return null;
+      });
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    // Start polling after we resolve the first supergraph
+    if (this.config.pollIntervalInMs) {
+      this.beginPolling();
+    }
+
+    if (!initialSupergraphSdl) {
+      throw new NoInitialSdlError();
+    }
+
+    this.completeSupergraphSdl = initialSupergraphSdl;
+
+    return {
+      cleanup: async (): Promise<void> => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          clearTimeout(this.timerRef);
+          this.timerRef = null;
+        }
+      },
+      supergraphSdl: initialSupergraphSdl,
+    };
+  }
+
+  private async updateSupergraphSdl(): Promise<string | null> {
+    if (!this.originalSubgraphs.length) {
+      throw new NoSubgraphsError();
+    }
+    const result = await loadServicesFromRemoteEndpoint({
+      getServiceIntrospectionHeaders: async service => {
+        return typeof this.config.introspectionHeaders === 'function'
+          ? // eslint-disable-next-line no-return-await
+            await this.config.introspectionHeaders(service)
+          : this.config.introspectionHeaders;
+      },
+      serviceList: this.originalSubgraphs,
+      serviceSdlCache: this.serviceSdlCache,
+    });
+
+    if (!result.isNewSchema) {
+      return this.completeSupergraphSdl;
+    }
+
+    if (!result.serviceDefinitions) {
+      return this.completeSupergraphSdl;
+    }
+
+    const { nodeTypeDefs } = result.serviceDefinitions.reduce<NodeTypes>(
+      (acc, service: ServiceDefinition) => {
+        const newServiceDefinition = visit(service.typeDefs, {
+          ObjectTypeDefinition(node) {
+            const name = node.name.value;
+            if (
+              acc.seenNodeTypes.has(name) ||
+              !IntrospectAddNodeAndCompose.objectImplementsNodeInterface(node)
+            ) {
+              return;
+            }
+            acc.nodeTypeDefs.push(
+              IntrospectAddNodeAndCompose.generateNodeType(name),
```

Comment:
I'd do this slightly different so it's more multi purpose (in case we need) and easier to read:
- collect all `name` that `objectImplementsNodeInterface`
- `map` the resulting array calling `generateNodeType`

this way the `reduce` becomes shorter and we can reuse the name of node-implementors for other stuff if we need, converting to the type definition when needed